### PR TITLE
[CA][Helm] Add ClusterAPI as supported Provider to the helm chart

### PR DESCRIPTION
**Pull request by MaxRink**
This PR add CAPI as a supported provider to the helm chart and orients itself on the existing config parameters.
It also adds an mode for running in a non-clusterscoped mode.

### CODE REVIEW

**Comment by detiber** [comment]
The cluster api related bits lgtm, but I'm not familiar enough with the helm templates to provide too much feedback there. maybe @gjtempleton can provide feedback there.

/lgtm

**Comment by MaxRink** [comment]
resolved the mergeconflict :)

**Comment by MaxRink** [comment]
Well, i have no idea about which comments ct lint actually complains

**Comment by dntosas** [comment]
cc @detiber any chance this one gets a review?

**Comment by gjtempleton** [comment]
/assign @gjtempleton

I'll try to give the chart-related pieces a look this week.

File: charts/cluster-autoscaler/values.yaml [hunk 1]
```diff
@@ -99,13 +104,29 @@ magnumClusterName: ""
 # magnumCABundlePath -- Path to the host's CA bundle, from `ca-file` in the cloud-config file.
 magnumCABundlePath: "/etc/kubernetes/ca-bundle.crt"
 
+# clusterAPIMode --  Cluster API mode, see https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/README.md#connecting-cluster-autoscaler-to-cluster-api-management-and-workload-clusters
+# Syntax: workloadClusterMode-ManagementClusterMode
+# for `kubeconfig-kubeconfig`, `incluster-kubeconfig` and `single-kubeconfig` you always must mount the external kubeconfig using either `extraVolumeSecrets`or `extraMounts` and `extraVolumes`
+# if you dont set `clusterAPIKubeconfigSecret`and thus use an in-cluster config or want to use a non capi generated kubeconfig you must do so for the workload kubeconfig as well
+clusterAPIMode: incluster-incluster  # incluster-incluster, incluster-kubeconfig, kubeconfig-incluster, kubeconfig-kubeconfig, single-kubeconfig
+
+# clusterAPIKubeconfigSecret -- Secret containing kubeconfig for connecting to Cluster API managed workloadcluster
```
Comment (gjtempleton) [inline charts/cluster-autoscaler/values.yaml:113]: From my reading of the deployment template, it's up to the user to provision this secret themselves, can we make that a bit clearer?

Maybe this is my unfamiliarity with ClusterAPI, but I don't seem to see workloadcluster as one word being a commonly used term?

File: charts/cluster-autoscaler/values.yaml [hunk 2]
```diff
@@ -3,24 +3,29 @@
 affinity: {}
 
 autoDiscovery:
-  # cloudProviders `aws`, `gce` and `magnum` are supported by auto-discovery at this time
+  # cloudProviders `aws`, `gce`, `magnum` and `clusterapi` are supported by auto-discovery at this time
   # AWS: Set tags as described in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md#auto-discovery-setup
 
   # autoDiscovery.clusterName -- Enable autodiscovery for `cloudProvider=aws`, for groups matching `autoDiscovery.tags`.
+  # autoDiscovery.clusterName -- Enable autodiscovery for `cloudProvider=clusterapi`, for groups matching `autoDiscovery.labels`.
```
Comment (gjtempleton) [inline charts/cluster-autoscaler/values.yaml:10]: For slightly neater readme instructions:
```suggestion
  # Enable autodiscovery for `cloudProvider=clusterapi`, for groups matching `autoDiscovery.labels`.
```

File: charts/cluster-autoscaler/values.yaml [hunk 3]
```diff
@@ -3,24 +3,29 @@
...
+    - k8s.io/cluster-autoscaler/{{ .Values.autoDiscovery.clusterName }}
   # - kubernetes.io/cluster/{{ .Values.autoDiscovery.clusterName }}
 
   # autoDiscovery.roles -- Magnum node group roles to match.
   roles:
-  - worker
+    - worker
 
+  # autoDiscovery.labels -- Cluster-API labels to match  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/README.md#configuring-node-group-auto-discovery
+  labels: {}
```
Comment (gjtempleton) [inline charts/cluster-autoscaler/values.yaml:26]:
```suggestion
  labels: []
```

File: charts/cluster-autoscaler/values.yaml [hunk 4]
```diff
@@ -99,13 +104,29 @@ magnumClusterName: ""
 # magnumCABundlePath -- Path to the host's CA bundle, from `ca-file` in the cloud-config file.
 magnumCABundlePath: "/etc/kubernetes/ca-bundle.crt"
 
+# clusterAPIMode --  Cluster API mode, see https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/README.md#connecting-cluster-autoscaler-to-cluster-api-management-and-workload-clusters
+# Syntax: workloadClusterMode-ManagementClusterMode
+# for `kubeconfig-kubeconfig`, `incluster-kubeconfig` and `single-kubeconfig` you always must mount the external kubeconfig using either `extraVolumeSecrets`or `extraMounts` and `extraVolumes`
```
Comment (gjtempleton) [inline charts/cluster-autoscaler/values.yaml:109]:
```suggestion
# for `kubeconfig-kubeconfig`, `incluster-kubeconfig` and `single-kubeconfig` you always must mount the external kubeconfig using either `extraVolumeSecrets` or `extraMounts` and `extraVolumes`
```

**Review by gjtempleton — COMMENTED** [review]
You'll need to rebase and update the chart version again.

Also some small queries and general comments, but nothing major so should be able to get this merged once those are addressed/answered.

File: charts/cluster-autoscaler/values.yaml (hunk 1 above)
Comment (MaxRink) [reply charts/cluster-autoscaler/values.yaml:113]: Its a common CAPI term
https://cluster-api.sigs.k8s.io/user/concepts.html#workload-cluster

**Comment by MaxRink** [comment]
@gjtempleton ive rebased and addressed the open points

**Comment by gjtempleton** [comment]
Thanks for that!
/lgtm
/approve

**Pull request merged**